### PR TITLE
Add options for books on kobo e-reader

### PR DESCRIPTION
**Pull request by typebrook**
## Add new arguments for e-reader

`--book_from <E-READER>` to specify connected e-reader (Only support "kobo" now)
`--device_path <PATH>` to specify path of mounting point

If these two options are specified, `obok.py` would check mounting point
of device, and show a list of available epub files in `/MOUNTING/POINT/.kobo/kepub`

User input the number of book need to be translated, then the decrypted epub
would be save to current directory. And this epub file would be pass to
`epub_loader`.

Usage would be like:
```bash
python make_book.py \                                                                                        1
  --model gpt3 \
  --language "Traditional Chinese" \
  --book_from kobo \
  --device_path /run/media/pham/KOBOeReader/
```

## Add `obok.py` as file selector for kobo device

Source: https://github.com/apprenticeharper/DeDRM_tools/blob/master/Obok_plugin/obok/obok.py

Since scripts in DeDRM_tools has no unified lisence. Directly copy
obok.py maybe a litte tricky. [^1]

So I mail to maintainer of DeDRM, asking for the permission.
Now waiting for his/her response.

[^1]: https://github.com/apprenticeharper/DeDRM_tools/issues/244

### CODE REVIEW

File: book_maker/obok.py [hunk 1]
```diff
@@ -0,0 +1,876 @@
...
+                                "%LOCALAPPDATA%"
+                            )
+                    if self.kobodir == "":
+                        if "USERPROFILE" in os.environ.keys():
+                            # Python 2.x does not return unicode env. Use Python 3.x
+                            self.kobodir = os.path.join(
+                                winreg.ExpandEnvironmentStrings("%USERPROFILE%"),
+                                "Local Settings",
+                                "Application Data",
+                            )
```
Comment (Arborym) [inline book_maker/obok.py:399]:
```suggestion
                    if (
                        sys.getwindowsversion().major > 5
                        and "LOCALAPPDATA" in os.environ.keys()
                    ):
                        # Python 2.x does not return unicode env. Use Python 3.x
                        self.kobodir = winreg.ExpandEnvironmentStrings(
                            "%LOCALAPPDATA%"
                        )
                    if self.kobodir == "" and "USERPROFILE" in os.environ.keys():
                        # Python 2.x does not return unicode env. Use Python 3.x
                        self.kobodir = os.path.join(
                            winreg.ExpandEnvironmentStrings("%USERPROFILE%"),
                            "Local Settings",
                            "Application Data",
                        )
```

File: book_maker/obok.py [hunk 2]
```diff
@@ -0,0 +1,876 @@
...
+                elif sys.platform.startswith("linux"):
+                    # sets ~/.config/calibre as the location to store the kobodir location info file and creates this directory if necessary
+                    kobodir_cache_dir = os.path.join(
+                        os.environ["HOME"], ".config", "calibre"
+                    )
+                    if not os.path.isdir(kobodir_cache_dir):
+                        os.mkdir(kobodir_cache_dir)
+
+                    # appends the name of the file we're storing the kobodir location info to the above path
+                    kobodir_cache_file = str(kobodir_cache_dir) + "/" + "kobo_location"
```
Comment (Arborym) [inline book_maker/obok.py:420]:
```suggestion
                    kobodir_cache_file = f"{str(kobodir_cache_dir)}/kobo_location"
```

File: book_maker/obok.py [hunk 3]
```diff
@@ -0,0 +1,876 @@
...
+            output = subprocess.Popen(
+                "wmic nic where PhysicalAdapter=True get MACAddress",
+                shell=True,
+                stdout=subprocess.PIPE,
+                text=True,
+            ).stdout
+            for line in output:
+                m = c.search(line)
+                if m:
+                    macaddrs.append(re.sub("-", ":", m.group(1)).upper())
```
Comment (Arborym) [inline book_maker/obok.py:552]:
```suggestion
                if m := c.search(line):
                    macaddrs.append(re.sub("-", ":", m[1]).upper())
```

File: book_maker/obok.py [hunk 4]
```diff
@@ -0,0 +1,876 @@
...
+            c = re.compile(
+                "\s(" + "[0-9a-f]{2}:" * 5 + "[0-9a-f]{2})(\s|$)", re.IGNORECASE
+            )
+            output = subprocess.check_output(
+                "/sbin/ifconfig -a", shell=True, encoding="utf-8"
+            )
+            matches = c.findall(output)
+            for m in matches:
+                # print "m:{0}".format(m[0])
+                macaddrs.append(m[0].upper())
```
Comment (Arborym) [inline book_maker/obok.py:563]:
```suggestion
            macaddrs.extend(m[0].upper() for m in matches)
```

File: book_maker/obok.py [hunk 5]
```diff
@@ -0,0 +1,876 @@
...
+            # probably linux
+
+            # let's try ip
+            c = re.compile(
+                "\s(" + "[0-9a-f]{2}:" * 5 + "[0-9a-f]{2})(\s|$)", re.IGNORECASE
+            )
+            for line in os.popen("ip -br link"):
+                m = c.search(line)
+                if m:
+                    macaddrs.append(m.group(1).upper())
```
Comment (Arborym) [inline book_maker/obok.py:574]:
```suggestion
                if m := c.search(line):
                    macaddrs.append(m[1].upper())
```

File: book_maker/obok.py [hunk 6]
```diff
@@ -0,0 +1,876 @@
...
+
+            # let's try ipconfig under wine
+            c = re.compile(
+                "\s(" + "[0-9a-f]{2}-" * 5 + "[0-9a-f]{2})(\s|$)", re.IGNORECASE
+            )
+            for line in os.popen("ipconfig /all"):
+                m = c.search(line)
+                if m:
+                    macaddrs.append(re.sub("-", ":", m.group(1)).upper())
```
Comment (Arborym) [inline book_maker/obok.py:583]:
```suggestion
                if m := c.search(line):
                    macaddrs.append(re.sub("-", ":", m[1]).upper())
```

File: book_maker/obok.py [hunk 7]
```diff
@@ -0,0 +1,876 @@
...
+
+    def __getuserids(self):
+        userids = []
+        cursor = self.__cursor.execute("SELECT UserID FROM user")
+        row = cursor.fetchone()
+        while row is not None:
+            try:
+                userid = row[0]
+                userids.append(userid)
+            except:
```
Comment (Arborym) [inline book_maker/obok.py:599]:
```suggestion
            except Exception:
```

File: book_maker/obok.py [hunk 8]
```diff
@@ -0,0 +1,876 @@
...
+                return True
+            else:
+                print("Bad XML: {0}".format(contents[:8]))
+                raise ValueError
+        elif self.mimetype == "image/jpeg":
+            if contents[:3] == b"\xff\xd8\xff":
+                return True
+            else:
+                print("Bad JPEG: {0}".format(contents[:3].hex()))
+                raise ValueError()
```
Comment (Arborym) [inline book_maker/obok.py:793]:
```suggestion
            print("Bad JPEG: {0}".format(contents[:3].hex()))
            raise ValueError()
```

File: book_maker/obok.py [hunk 9]
```diff
@@ -0,0 +1,876 @@
...
+        """
+        Remove the trailing padding, using what appears to be the CMS
+        algorithm from RFC 5652 6.3"""
+        lastchar = binascii.b2a_hex(contents[-1:])
+        strlen = int(lastchar, 16)
+        padding = strlen
+        if strlen == 1:
+            return contents[:-1]
+        if strlen < 16:
+            for i in range(strlen):
```
Comment (Arborym) [inline book_maker/obok.py:806]:
```suggestion
            for _ in range(strlen):
```

File: book_maker/obok.py [hunk 10]
```diff
@@ -0,0 +1,876 @@
...
+            elif (
+                contents[:22]
+                == b"\xff\xfe<\x00!\x00D\x00O\x00C\x00T\x00Y\x00P\x00E\x00"
+            ):
+                # utf-16LE of weird <!DOCTYPE start
+                return True
+            else:
+                print("Bad XML: {0}".format(contents[:8]))
+                raise ValueError
+        elif self.mimetype == "image/jpeg":
```
Comment (Arborym) [inline book_maker/obok.py:788]:
```suggestion
        if self.mimetype == "image/jpeg":
```

**Review by Arborym — CHANGES_REQUESTED** [review]
Refactoring/Optimizations

File: book_maker/cli.py [hunk 11]
```diff
@@ -114,6 +129,14 @@ def main():
     else:
         OPENAI_API_KEY = ""
 
+    if options.book_from == "kobo":
+        device_path = options.device_path
+        if device_path == None:
```
[the remark this answers is not in the document]
Comment (typebrook) [reply book_maker/cli.py]: Resolved in 3748b83

File: book_maker/obok.py [hunk 12]
```diff
@@ -0,0 +1,876 @@
...
+    AES_ecb_encrypt = F(None, "AES_ecb_encrypt", [c_char_p, c_char_p, AES_KEY_p, c_int])
+
+    class AES(object):
+        def __init__(self, userkey):
+            self._blocksize = len(userkey)
+            if (
+                (self._blocksize != 16)
+                and (self._blocksize != 24)
+                and (self._blocksize != 32)
+            ):
```
[the remark this answers is not in the document]
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py [hunk 13]
```diff
@@ -0,0 +1,876 @@
...
+
+
+# Wrap a stream so that output gets flushed immediately
+# and also make sure that any unicode strings get
+# encoded using "replace" before writing them.
+class SafeUnbuffered:
+    def __init__(self, stream):
+        self.stream = stream
+        self.encoding = stream.encoding
+        if self.encoding == None:
```
[the remark this answers is not in the document]
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py [hunk 14]
```diff
@@ -0,0 +1,876 @@
...
+
+
+class KoboLibrary(object):
+    """The Kobo library.
+
+    This class represents all the information available from the data
+    written by the Kobo Desktop Edition application, including the list
+    of books, their titles, and the user's encryption key(s)."""
+
+    def __init__(self, serials=[], device_path=None, desktopkobodir=""):
```
[the remark this answers is not in the document]
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py [hunk 15]
```diff
@@ -0,0 +1,876 @@
...
+                            if "deviceSerial" in node.tag:
+                                serial = node.text
+                                # print "found serial {0}".format(serial)
+                                serials.append(serial)
+                                break
+                    else:
+                        # print "cannot get serials from device."
+                        device_path = ""
+                        self.kobodir = ""
+                        kobodb = ""
```
[the remark this answers is not in the document]
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 3 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 4 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 5 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 6 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py [hunk 16]
```diff
@@ -0,0 +1,876 @@
...
+        ocf = ET.fromstring(zin.read("META-INF/container.xml"))
+        opffile = ocf.find(".//ocf:rootfile", xmlns).attrib["full-path"]
+        basedir = re.sub("[^/]+$", "", opffile)
+        opf = ET.fromstring(zin.read(opffile))
+        zin.close()
+
+        c = re.compile("/")
+        for item in opf.findall(".//opf:item", xmlns):
+            mimetype = item.attrib["media-type"]
+
```
[the remark this answers is not in the document]
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py [hunk 17]
```diff
@@ -0,0 +1,876 @@
...
+            # Convert relative URIs
+            href = item.attrib["href"]
+            if not c.match(href):
+                href = "".join((basedir, href))
+
+            # Update books we've found from the DB.
+            if href in self._encryptedfiles:
```
[the remark this answers is not in the document]
Comment (typebrook) [reply book_maker/obok.py:674]: Resolved in 3748b83

File: book_maker/obok.py [hunk 18]
```diff
@@ -0,0 +1,876 @@
...
+                self._encryptedfiles[href].mimetype = mimetype
+        return self._encryptedfiles
+
+    @property
+    def has_drm(self):
+        return not self.type == "drm-free"
```
[the remark this answers is not in the document]
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 8 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 10 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 9 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 7 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 1 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

File: book_maker/obok.py (hunk 2 above)
Comment (typebrook) [reply book_maker/obok.py]: Resolved in 3748b83

**Comment by typebrook** [comment]
@Arborym Python ninja huh? Please tell me which syntax checker you are using. (I have few experience about coding in Python)

File: book_maker/obok.py [hunk 19]
```diff
@@ -0,0 +1,865 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+# Version 4.1.1 March 2023
+# Make obok.py works as file selector
+
+# Version 4.1.0 February 2021
+# Add detection for Kobo directory location on Linux
```
Comment (yihong0618) [inline book_maker/obok.py:7]: Comments in the first its a vender file.

**Review by yihong0618 — APPROVED** [review]

File: book_maker/obok.py (hunk 19 above)
Comment (typebrook) [reply book_maker/obok.py:7]: Sorry, I don't understand. You want me to remove shebang?

Comment (yihong0618) [reply book_maker/obok.py:7]: sorry what I mean since its a vender file
We can add something like

```suggestion
# the original code from …
```

Comment (typebrook) [reply book_maker/obok.py:7]: Good idea!

Comment (typebrook) [reply book_maker/obok.py]: Resolved in e5b5687

**Comment by yihong0618** [comment]
@typebrook can you fix the conflicts?>

**Comment by typebrook** [comment]
Sure

**Comment by yihong0618** [comment]
Thanks!

**Comment by Arborym** [comment]
> @Arborym Python ninja huh? Please tell me which syntax checker you are using. (I have few experience about coding in Python)

I just use standard Pylance and Pylint in VS Code 🙂

**Pull request merged**